src/Model/ModelRegistry.php

```diff
@@ -82,13 +87,36 @@ public function __construct($modelDescribers, OA\OpenApi $api, array $alternativ
     public function register(Model $model): string
     {
         $hash = $model->getHash();
-        if (!isset($this->models[$hash])) {
-            $this->models[$hash] = $model;
-            $this->unregistered[] = $hash;
-        }
+        $type = $this->getTypeShortName($model->getType());
+
+        $schema = null;
         if (!isset($this->names[$hash])) {
             $this->names[$hash] = $this->generateModelName($model);
             $this->registeredModelNames[$this->names[$hash]] = $model;
+
+            $schema = $this->getSchemaWithoutRegistration($model);
+
+            // Only try to match schemas if we successfully got one
+            if (null !== $schema) {
+                foreach ($this->schemasPerType[$type] ?? [] as $schemaAndModel) {
+                    if ($schemaAndModel['schema'] === json_encode($schema->jsonSerialize())) {
+                        $newHash = $schemaAndModel['model']->getHash();
+                        unset($this->names[$hash], $this->registeredModelNames[$hash]);
+
+                        return OA\Components::SCHEMA_REF.$this->names[$newHash];
+                    }
+                }
+            }
+        }
+
+        if (!isset($this->models[$hash])) {
+            $this->models[$hash] = $model;
+            $this->unregistered[] = $hash;
+            $this->unregistered = array_unique($this->unregistered);
+            // Only store schema if it was successfully generated
+            if (null !== $schema) {
+                $this->schemasPerType[$type][] = ['schema' => json_encode($schema), 'model' => $model];
```

Comment:
Can we not store the Schema object directly instead of json encoding it?
```suggestion
                $this->schemasPerType[$type][] = ['schema' => $schema, 'model' => $model];
```

Reply:
We can do that yes, but we have to use json_encode for comparison. Because:
- We cannot use strict equality here as instances are different
- We cannot use lose equality either, because of a cyclic dependency we go into an infinite loop

So if we decide to store the schema and not it's json representation, we would have to run json_encode and there is a lot of ops in `OpenApi\Annotations\AbstractAnnotation::jsonSerialize`, so I assumed it would be better to json_encode only once.
WDYT ?

Reply:
only once is fine then :)